### PR TITLE
Optimize functions `mask_landsea()`, `mask_landseaice()` and `calculate_volume()` for lazy input

### LINKED ISSUE #2514: Poor performance of lazy `mask_landsea` preprocessor
**Describe the bug**

The preprocessor `mask_landsea` is performing poorly with high-res data. When running the following recipe with dask distributed, the task execution basically freezes. Removing the masking preprocessor fixes this issue.

Example recipe:

```yml
documentation:
  title: High-res evaluation
  description: Simple evaluation of high-res model output.
  authors:
    - schlund_manuel

preprocessors:
  nh_land_mean_annual_cycle:
    mask_landsea:
      mask_out: sea
    extract_region:
      start_latitude: 30.0
      end_latitude: 90.0
      start_longitude: 0.0
      end_longitude: 360.0
    area_statistics:
      operator: mean
    monthly_statistics:
      operator: mean
    climate_statistics:
      operator: mean
      period: month

diagnostics:
  annual_cycle:
    variables:
      tas:
        mip: day
        project: CMIP6
        exp: highresSST-present
        preprocessor: nh_land_mean_annual_cycle
        timerange: 1950/1960
    scripts: null

datasets:
  - {dataset: NICAM16-9S, ensemble: r1i1p1f1, grid: gr}
```

I could trace that to the usage of `da` instead `np` [here](https://github.com/ESMValGroup/ESMValCore/blob/412ac53727e8e545d78bb6f98621e6fef9149911/esmvalcore/preprocessor/_mask.py#L288). In addition, the [broadcasting](https://github.com/ESMValGroup/ESMValCore/blob/412ac53727e8e545d78bb6f98621e6fef9149911/esmvalcore/preprocessor/_mask.py#L317) actually never happens since the function's return value is not used.

I will provide a fix for that.

### PR DESCRIPTION
**Pull request by schlunma**
## Description

This PR
- uses the newly added `chunks` for `iris.util.broadcast_to_shape` whenever possible in our code base to improve lazy computations.
- fixes a small bug for `mask_landsea` (see https://github.com/ESMValGroup/ESMValCore/issues/2514 for details).
- makes sure that the functions `mask_landsea()`, `mask_landseaice()` and `calculate_volume()` return lazy data for lazy input, but also realized data if all input is realized.

Closes #2514

***

## [Before you get started](https://docs.esmvaltool.org/projects/ESMValCore/en/latest/contributing.html#getting-started)

- [x] [☝ Create an issue](https://github.com/ESMValGroup/ESMValCore/issues) to discuss what you are going to do

## [Checklist](https://docs.esmvaltool.org/projects/ESMValCore/en/latest/contributing.html#checklist-for-pull-requests)

It is the responsibility of the author to make sure the pull request is ready to review. The icons indicate whether the item will be subject to the [🛠 Technical][1] or [🧪 Scientific][2] review.

[1]: https://docs.esmvaltool.org/en/latest/community/review.html#technical-review
[2]: https://docs.esmvaltool.org/en/latest/community/review.html#scientific-review

- [x] [🧪][2] The new functionality is [relevant and scientifically sound](https://docs.esmvaltool.org/projects/ESMValCore/en/latest/contributing.html#scientific-relevance)
- [x] [🛠][1] This pull request has a [descriptive title and labels](https://docs.esmvaltool.org/projects/ESMValCore/en/latest/contributing.html#pull-request-title-and-label)
- [x] [🛠][1] Code is written according to the [code quality guidelines](https://docs.esmvaltool.org/projects/ESMValCore/en/latest/contributing.html#code-quality)
- [x] [🧪][2] and [🛠][1] [Documentation](https://docs.esmvaltool.org/projects/ESMValCore/en/latest/contributing.html#documentation) is available
- [x] [🛠][1] [Unit tests](https://docs.esmvaltool.org/projects/ESMValCore/en/latest/contributing.html#tests) have been added
- [x] [🛠][1] Changes are [backward compatible](https://docs.esmvaltool.org/projects/ESMValCore/en/latest/contributing.html#backward-compatibility)
- [x] [🛠][1] Any changed [dependencies have been added or removed](https://docs.esmvaltool.org/projects/ESMValCore/en/latest/contributing.html#dependencies) correctly
- [x] [🛠][1] The [list of authors](https://docs.esmvaltool.org/projects/ESMValCore/en/latest/contributing.html#list-of-authors) is up to date
- [x] [🛠][1] All [checks below this pull request](https://docs.esmvaltool.org/projects/ESMValCore/en/latest/contributing.html#pull-request-checks) were successful

***

To help with the number pull requests:

- 🙏 We kindly ask you to [review](https://docs.esmvaltool.org/en/latest/community/review.html#review-of-pull-requests) two other [open pull requests](https://github.com/ESMValGroup/ESMValCore/pulls) in this repository

### CODE REVIEW

**Review by bouweandela — CHANGES_REQUESTED** [review]
> makes sure that the functions `mask_landsea()`, `mask_landseaice()` and `calculate_volume()` return lazy data for lazy input, but also realized data for realized input.

The question here is how to treat a mix of input arrays. The common choice seems to be that the output is lazy if any of the input arrays are lazy, e.g. `np.ones(2) * da.ones(2)` results in a Dask array. This kind of makes sense because it provides the lowest memory footprint and still offers the user full control because they can make all input arrays realized if they want realized output. Would that work for you?

**Comment by schlunma** [comment]
> The question here is how to treat a mix of input arrays. The common choice seems to be that the output is lazy if any of the input arrays are lazy, e.g. `np.ones(2) * da.ones(2)` results in a Dask array. This kind of makes sense because it provides the lowest memory footprint and still offers the user full control because they can make all input arrays realized if they want realized output. Would that work for you?

Yes, that makes total sense, and also has the benefit of easier code. I chose to make it only depend on the actual data, not on the ancillary variables. I will change that now 👍

File: esmvalcore/preprocessor/_mask.py [hunk 1]
```diff
@@ -313,13 +321,22 @@ def _mask_with_shp(cube, shapefilename, region_indices=None):
...
 
-    mask = da.array(mask)
-    iris.util.broadcast_to_shape(mask, cube.shape, cube.coord_dims('latitude')
-                                 + cube.coord_dims('longitude'))
+    if cube.has_lazy_data():
+        mask = da.array(mask)
+        chunks = cube.lazy_data().chunks
+    else:
+        chunks = None
+    mask = iris.util.broadcast_to_shape(
```
[the remark this answers is not in the document]
Comment (schlunma) [reply esmvalcore/preprocessor/_mask.py:330]: Done in https://github.com/ESMValGroup/ESMValCore/pull/2515/commits/0178f40b1edc1626d536530e72c3486e4a0a99e6

File: esmvalcore/preprocessor/_volume.py [hunk 2]
```diff
@@ -152,9 +152,7 @@ def calculate_volume(cube: Cube) -> da.core.Array:
             f'Cannot compute volume using the Z-axis. {err}') from err
 
     # Calculate Z-direction thickness
-    thickness = depth.core_bounds()[..., 1] - depth.core_bounds()[..., 0]
-    if cube.has_lazy_data():
-        thickness = da.array(thickness)
```
[the remark this answers is not in the document]
Comment (schlunma) [reply esmvalcore/preprocessor/_volume.py]: Done in https://github.com/ESMValGroup/ESMValCore/pull/2515/commits/74a8aeb88ef6b0a7e50d5b139566cf4acdf16bc0

**Comment by schlunma** [comment]
Thanks for the review Bouwe! I implemented all your suggestinos (also for the masking preprocessor [here](https://github.com/ESMValGroup/ESMValCore/pull/2515/commits/c8d6e0933e279bd59e13d488f5ac3a63d8a1d733)). Could you please have another look?

File: esmvalcore/preprocessor/_mask.py [hunk 3]
```diff
@@ -119,28 +127,28 @@ def mask_landsea(cube, mask_out):
...
+            )
 
     if fx_cube:
-        fx_cube_data = da.broadcast_to(fx_cube.core_data(), cube.shape)
-        landsea_mask = _get_fx_mask(fx_cube_data, mask_out,
-                                    fx_cube.var_name)
-        cube.data = _apply_fx_mask(landsea_mask, cube.core_data())
+        fx_cube_data = np.broadcast_to(fx_cube.core_data(), cube.shape)
+        landsea_mask = _get_fx_mask(fx_cube_data, mask_out, fx_cube.var_name)
+        cube.data = _apply_mask(landsea_mask, cube.core_data())
```
Comment (bouweandela) [inline esmvalcore/preprocessor/_mask.py:138]: Maybe the `broadcast_to` operation could be moved into `_apply_mask` and then be supplied with a `chunks` argument so it uses the chunks that are aligned with the cube's data (perhaps by using `iris.util.broadcast_to_shape`)? Calling the function `_get_fx_mask` before broadcasting may be a bit more computationally efficient too.

**Review by bouweandela — CHANGES_REQUESTED** [review]
Nice work @schlunma! The type hints really make this code a lot more readable. A few more suggestions.

File: esmvalcore/preprocessor/_mask.py (hunk 3 above)
Comment (schlunma) [reply esmvalcore/preprocessor/_mask.py]: Great suggestion! Done inhttps://github.com/ESMValGroup/ESMValCore/pull/2515/commits/325db6db180f5249ecd7bc0d120419e95cea70b3

File: esmvalcore/preprocessor/_mask.py [hunk 4]
```diff
@@ -50,22 +60,29 @@ def _get_fx_mask(fx_data, fx_option, mask_type):
...
-
-    return var_data
+def _apply_mask(
+    mask: np.ndarray | da.Array,
+    array: np.ndarray | da.Array,
+    dim_map: Optional[Iterable[int]] = None,
+) -> np.ndarray | da.Array:
+    """Apply a (broadcasted) mask on an array."""
+    npx = get_array_module(mask, array)
+    if dim_map is not None:
```
Comment (bouweandela) [inline esmvalcore/preprocessor/_mask.py:70]: Why is this condition needed? The function appears to be always called with a `dim_map` argument.

File: esmvalcore/preprocessor/_mask.py (hunk 3 above)
Comment (bouweandela) [reply esmvalcore/preprocessor/_mask.py]: :partying_face:

File: esmvalcore/preprocessor/_mask.py (hunk 4 above)
Comment (schlunma) [reply esmvalcore/preprocessor/_mask.py:70]: The `tests` use that function without a `dim_map`. I think it's nice to be able to call this function without broadcasting.

Comment (bouweandela) [reply esmvalcore/preprocessor/_mask.py:70]: If the feature is not used by any non-test code I would prefer to remove it, as it makes it more complicated to understand the code at no current benefit. If there is a use case in the future we could always add it again.

**Review by bouweandela — APPROVED** [review]
Thanks for these improvements @schlunma, good to see these preprocessor functions getting some love.

File: esmvalcore/preprocessor/_mask.py (hunk 4 above)
Comment (bouweandela) [reply esmvalcore/preprocessor/_mask.py:70]: Anyway, the code is in a much better shape with the changes in this pull request than it was before, so I'm happy to see this merged regardless of this point.

**Pull request merged**